File: common/src/main/scala/com/taskadapter/connector/definition/SaveResultBuilder.scala
package com.taskadapter.connector.definition

import com.taskadapter.model.GTask

class SaveResultBuilder {

  private var targetFileAbsolutePath = ""
  /**
    * Number of updated tasks.
    */
  private var updatedTasksNumber = 0
  /**
    * Number of created tasks.
    */
  private var createdTasksNumber = 0

  // maps original task Key --> new task info when new tasks are created
  private val idToRemoteKeyMap = new scala.collection.mutable.ListBuffer[TaskKeyMapping]
  private val generalErrors = new scala.collection.mutable.ListBuffer[Throwable]
  private val taskErrors = new scala.collection.mutable.ListBuffer[TaskError]

  def setTargetFileAbsolutePath(targetFileAbsolutePath: String): Unit = {
    this.targetFileAbsolutePath = targetFileAbsolutePath
  }

  def addCreatedTask(original: TaskId, newKey: TaskId): Unit = {
    idToRemoteKeyMap += new TaskKeyMapping(original, newKey);
    createdTasksNumber += 1
  }

  def addUpdatedTask(original: TaskId, newId: TaskId): Unit = {
    idToRemoteKeyMap += new TaskKeyMapping(original, newId)
    updatedTasksNumber += 1
  }

  def getRemoteKey(original: TaskId): TaskId = idToRemoteKeyMap.find(long => long.originalId == original).map(e => e.newId).orNull

  def addGeneralError(e: Throwable): Unit = {
    generalErrors += e
  }

  def addTaskError(task: GTask, e: Exception): Unit = {
    taskErrors += new TaskError(task, e)
  }

  def getResult = SaveResult(targetFileAbsolutePath, updatedTasksNumber, createdTasksNumber,
    idToRemoteKeyMap.toList,
    generalErrors.toList,
    taskErrors.toList
  )
}

File: common/src/main/java/com/taskadapter/connector/definition/SaveResultBuilder.java
package com.taskadapter.connector.definition;

import com.taskadapter.model.GTask;

import java.util.ArrayList;
import java.util.List;

public class SaveResultBuilder {

    private String targetFileAbsolutePath = "";
    /**
     * Number of updated tasks.
     */
    private int updatedTasksNumber = 0;
    /**
     * Number of created tasks.
     */
    private int createdTasksNumber = 0;

    // maps original task Key --> new task info when new tasks are created
    private final List<TaskKeyMapping> idToRemoteKeyMap = new ArrayList<>();
    private final List<Throwable> generalErrors = new ArrayList<>();
    private final List<TaskError> taskErrors = new ArrayList<>();

    public void setTargetFileAbsolutePath(String targetFileAbsolutePath) {
        this.targetFileAbsolutePath = targetFileAbsolutePath;
    }

    public void addCreatedTask(TaskId original, TaskId newKey) {
        idToRemoteKeyMap.add(new TaskKeyMapping(original, newKey));
        createdTasksNumber += 1;
    }

    public void addUpdatedTask(TaskId original, TaskId newId) {
        idToRemoteKeyMap.add(new TaskKeyMapping(original, newId));
        updatedTasksNumber += 1;
    }

    public TaskId getRemoteKey(TaskId original) {
        return idToRemoteKeyMap.stream()
                .filter(lon -> lon.originalId.equals(original))
                .map(e -> e.newId)
                .findFirst()
                .orElse(null);
    }

    public void addGeneralError(Throwable e) {
        generalErrors.add(e);
    }

    public void addTaskError(GTask task, Exception e) {
        taskErrors.add(new TaskError(task, e));
    }

    public SaveResult getResult() {
        return new SaveResult(targetFileAbsolutePath, updatedTasksNumber, createdTasksNumber,
                idToRemoteKeyMap,
                generalErrors,
                taskErrors
        );
    }
}
